server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -476,73 +464,106 @@ protected MapperBuilderContext createChildContext(MapperBuilderContext mapperBui
+            if (existing instanceof NestedObjectMapper == false && mergeWith instanceof NestedObjectMapper) {
+                // TODO stop NestedObjectMapper extending ObjectMapper?
+                MapperErrors.throwNestedMappingConflictError(mergeWith.name());
             }
-        }
-
-        MapperBuilderContext objectBuilderContext = createChildContext(parentBuilderContext, simpleName());
-        Map<String, Mapper> mergedMappers = null;
-        for (Mapper mergeWithMapper : mergeWith) {
-            Mapper mergeIntoMapper = (mergedMappers == null ? mappers : mergedMappers).get(mergeWithMapper.simpleName());
-
-            Mapper merged;
-            if (mergeIntoMapper == null) {
-                merged = mergeWithMapper;
-            } else if (mergeIntoMapper instanceof ObjectMapper objectMapper) {
-                merged = objectMapper.merge(mergeWithMapper, reason, objectBuilderContext);
+            ObjectMapper mergeWithObject = (ObjectMapper) mergeWith;
+            final Explicit<Boolean> enabled;
+            if (mergeWithObject.enabled.explicit()) {
+                if (reason == MergeReason.INDEX_TEMPLATE) {
+                    enabled = mergeWithObject.enabled;
+                } else if (existing.isEnabled() != mergeWithObject.isEnabled()) {
+                    throw new MapperException("the [enabled] parameter can't be updated for the object mapping [" + existing.name() + "]");
+                } else {
+                    enabled = existing.enabled;
+                }
             } else {
-                assert mergeIntoMapper instanceof FieldMapper || mergeIntoMapper instanceof FieldAliasMapper;
-                if (mergeWithMapper instanceof NestedObjectMapper) {
-                    MapperErrors.throwNestedMappingConflictError(mergeWithMapper.name());
-                } else if (mergeWithMapper instanceof ObjectMapper) {
-                    MapperErrors.throwObjectMappingConflictError(mergeWithMapper.name());
+                enabled = existing.enabled;
+            }
+            final Explicit<Boolean> subObjects;
+            if (mergeWithObject.subobjects.explicit()) {
+                if (reason == MergeReason.INDEX_TEMPLATE) {
+                    subObjects = mergeWithObject.subobjects;
+                } else if (existing.subobjects != mergeWithObject.subobjects) {
+                    throw new MapperException(
+                        "the [subobjects] parameter can't be updated for the object mapping [" + existing.name() + "]"
+                    );
+                } else {
+                    subObjects = existing.subobjects;
                 }
+            } else {
+                subObjects = existing.subobjects;
+            }
+            MapperBuilderContext objectBuilderContext = existing.createChildContext(parentBuilderContext, existing.simpleName());
+            Map<String, Mapper> mergedMappers = null;
```

Comment:
It feels a bit confusing to me to have this set to `null` and then maybe set up as part of the loop below.  It might be a bit clearer if we extracted this into another method, which could then return fast if either the input map or the to-merge map are empty.